### PR TITLE
Use CPI as common library in CAPVCD

### PR DESCRIPTION
**Pull request by Anirudh9794**
Ticket: VCDA-3524

* Remove vcdclient in CAPVCD
* Remove vcdswaggerclient in CAPVCD
* Refer to CPI main branch for the common library

Testing:
* Check if the code compiles properly
* cluster create
* cluster resize
* cluster delete

Signed-off-by: Aniruddha Shamasundar <aniruddha.9794@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131)

### CODE REVIEW

**Review by sahithi — APPROVED** [review]
:lgtm: with few comments

> Reviewed 107 of 107 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-:-N1ojsM058RGRobm9RwT:b-qo2juh)* status: all files reviewed, 3 unresolved discussions (waiting on @Anirudh9794)

___
*[`controllers/vcdcluster_controller.go` line 688 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1ojVsF1UaABKGOp9lX:-N1ojVsF1UaABKGOp9lY:b-7b0s1s) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/controllers/vcdcluster_controller.go#L688)):*
> ```Go
> 	}
> 
> 	gateway, err := capisdk.NewCapvcdGatewayManager(ctx, workloadVCDClient, vcdCluster.Spec.OvdcNetwork, r.Config.VCD.VIPSubnet)
> ```

as mentioned in another comment, let us aim for a single GatewayManager.
___
*[`pkg/capisdk/capi_gateway_manager.go` line 24 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1oiZuu8h9kpxTj5-sr:-N1oiZuu8h9kpxTj5-ss:br5plp2) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/pkg/capisdk/capi_gateway_manager.go#L24)):*
> ```Go
> }
> 
> func (capvcdGatewayManager *CapvcdGatewayManager) CreateL4LoadBalancer(ctx context.Context, virtualServiceNamePrefix string,
> ```

Aniruddh, As discussed before, we need to move these methods as well in CPI. Please create a task in the CAPVCD backlog if not already created.
CreateL4LoadBalancer
Update*
Delete*

After 3 months, we will all forget all the places we will need to fix a bug in the gateway manager code. We don't want to ship 3 different products for one fix.
___
*[`pkg/capisdk/defined_entity.go` line 21 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1oj2CUAaCQlD4X1Fiw:-N1oj2CUAaCQlD4X1Fix:b-i3rjdx) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/pkg/capisdk/defined_entity.go#L21)):*
> ```Go
> )
> 
> type CapvcdRdeManager struct {
> ```

RDE manager is the only file that we can have per component. Ideally, this should also go to CPI.

we can leave this as-is, but please do create a task with a detailed description to reevaluate the situation in a few weeks.
We will have to periodically visit and reevaluate how these things need to be organized.

**Review by Anirudh9794 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131)* status: 100 of 107 files reviewed, 3 unresolved discussions (waiting on @sahithi)

___
*[`controllers/vcdcluster_controller.go` line 688 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1ojVsF1UaABKGOp9lX:-N1poYDdAHiwGPvf4NR3:b-ldl8v3) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/controllers/vcdcluster_controller.go#L688)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

as mentioned in another comment, let us aim for a single GatewayManager.
</blockquote></details>

Created VCDA-3680
___
*[`pkg/capisdk/capi_gateway_manager.go` line 24 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1oiZuu8h9kpxTj5-sr:-N1pOmhS39W0VIF2CwFW:b-ldl8v3) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/pkg/capisdk/capi_gateway_manager.go#L24)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Aniruddh, As discussed before, we need to move these methods as well in CPI. Please create a task in the CAPVCD backlog if not already created.
CreateL4LoadBalancer
Update*
Delete*

After 3 months, we will all forget all the places we will need to fix a bug in the gateway manager code. We don't want to ship 3 different products for one fix.
</blockquote></details>

Created VCDA-3680
___
*[`pkg/capisdk/defined_entity.go` line 21 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1oj2CUAaCQlD4X1Fiw:-N1poe0k6WNo9sjjQxqW:b-ndoyc0) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/bcfc3233c7cc586c2f7880895bf4c322462bc5ec/pkg/capisdk/defined_entity.go#L21)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

RDE manager is the only file that we can have per component. Ideally, this should also go to CPI.

we can leave this as-is, but please do create a task with a detailed description to reevaluate the situation in a few weeks.
We will have to periodically visit and reevaluate how these things need to be organized.
</blockquote></details>

This file contains functions to modify only CAPVCD RDE and uses structs for RDE which are specific to CAPVCD. Moving this to vcdsdk in CPI will break our design that we should not define structs for one component in another.

File: pkg/capisdk/capi_gateway_manager.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	return &CapvcdGatewayManager{
+		gatewayManager: gatewayManager,
+	}, nil
+}
+
+func (capvcdGatewayManager *CapvcdGatewayManager) CreateL4LoadBalancer(ctx context.Context, virtualServiceNamePrefix string,
+	lbPoolNamePrefix string, ips []string, tcpPort int32, externalTcpPort int32, oneArm *vcdsdk.OneArm) (string, error) {
+
+	if capvcdGatewayManager.gatewayManager == nil {
+		return "", fmt.Errorf("referece to gateway manager is nil")
```
Comment (arunmk) [inline pkg/capisdk/capi_gateway_manager.go:28]: typo referece

File: pkg/capisdk/capi_gateway_manager.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	gatewayManager, err := vcdsdk.NewGatewayManager(ctx, client, networkName, ipamSubnet)
+	if err != nil {
+		return nil, fmt.Errorf("error creating gateway manager: [%v]", err)
+	}
+	return &CapvcdGatewayManager{
+		gatewayManager: gatewayManager,
+	}, nil
+}
+
+func (capvcdGatewayManager *CapvcdGatewayManager) CreateL4LoadBalancer(ctx context.Context, virtualServiceNamePrefix string,
```
Comment (arunmk) [inline pkg/capisdk/capi_gateway_manager.go:24]: Can't we merge this functionality into the HTTP code itself? There is already L4 getting handled along with HTTP/HTTPS in CPI code, so potentially this function could be removed.

File: pkg/capisdk/capi_gateway_manager.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	klog.V(3).Infof("Using external IP [%s] for virtual service\n", externalIP)
+
+	for _, portDetail := range portDetails {
+		if portDetail.internalPort == 0 {
+			klog.V(3).Infof("No internal port specified for [%s], hence loadbalancer not created\n", portDetail.portSuffix)
+			continue
+		}
+
+		virtualServiceName := fmt.Sprintf("%s-%s", virtualServiceNamePrefix, portDetail.portSuffix)
```
Comment (arunmk) [inline pkg/capisdk/capi_gateway_manager.go:75]: same comment about converting these to functions

**Review by arunmk — APPROVED** [review]
I am fine with things mostly. There are many comments and some of them can be ignored while some can be covered in other tickets.

The core issue that I want covered in this ticket is the userOrg issue.

The core issue which at least needs another ticket (or can be fixed here) is the removal of CreateL4LB function.

**Review by Anirudh9794 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131)* status: 99 of 107 files reviewed, 13 unresolved discussions (waiting on @arunmk and @sahithi)

___
*[`controllers/vcdcluster_controller.go` line 573 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-573:-N1rSihi-X9-JLMZ0Wzu:b-gmw376) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L573)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

when can we get `err == nil && controlPlaneNodeIP == ""` ?
</blockquote></details>

In the common code, the GetLoadBalancer() call is like 
```
// GetLoadBalancer :
func (gatewayManager *GatewayManager) GetLoadBalancer(ctx context.Context, virtualServiceName string, oneArm *OneArm) (string, error) {

	vsSummary, err := gatewayManager.GetVirtualService(ctx, virtualServiceName)
	if err != nil {
		return "", fmt.Errorf("unable to get summary for LB Virtual Service [%s]: [%v]",
			virtualServiceName, err)
	}
	if vsSummary == nil {
		return "", nil // this is not an error
	}
```
so if vsSummary is not found, there will be a case where `err == nil && controlPlaneNodeIP == ""`
___
*[`controllers/vcdcluster_controller.go` line 608 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-608:-N1rU3Ww4vlIB9iyN1au:b-9b0ccu) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L608)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

This `NoRdePrefix` should also idealy come from the common code
</blockquote></details>

Created follow up task VCDA-3682
___
*[`controllers/vcdcluster_controller.go` line 680 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-680:-N1soM8z0m5KAae_ge7S:b-okm0d5) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L680)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

this should be userorg in the second usage, correct? We should run through all such occurrences and clean them up. We should also use a common function to spit out the userorg and username
</blockquote></details>

There is already a function in the common library to spit out userorg and username
___
*[`controllers/vcdcluster_controller.go` line 695 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-695:-N1sovad-bHhcup8O70R:bua3l9p) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L695)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

My first thought was: We need to add helpers to construct these names as well and they should lie in CPI. We can create a new ticket and get it handled by someone else.

But second thought is that the client should specify their own names and construct LBs as they wish.

So my final thoughts are that construction of the name should become a function here, but it should be local to capvcd. The reason for it to be a function is to just ensure that create and delete use the same name.
</blockquote></details>

Created functions `GetVirtualServiceNamePrefix` and `GetLoadBalancerPoolNamePrefix`
___
*[`controllers/vcdcluster_controller.go` line 696 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-696:-N1sp02zCOFcV9iAtno3:bua3l9p) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L696)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

This one should be behind the same function as above
</blockquote></details>

Created functions `GetVirtualServiceNamePrefix` and `GetLoadBalancerPoolNamePrefix`
___
*[`controllers/vcdcluster_controller.go` line 697 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L0-r2-697:-N1sqSIj0F-l85sFqUpA:bm9qi22) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdcluster_controller.go#L697)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

Nit: passing the oneArm during delete is not a clean pattern, but we can live with it.
</blockquote></details>

Updated VCDA-3682 to track the removal of oneArm from DeleteLoadBalancer()
___
*[`controllers/vcdmachine_controller.go` line 552 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1Grh9sWbakcJFx24L1-r2-552:-N1sqmku9aV-EHGuXR9I:b-qzrf5g) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/controllers/vcdmachine_controller.go#L552)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

vapp name and machine name are swapped. I hope they are swapped in the called function as well.
</blockquote></details>

Ah, sorry. The first parameter should have been the VM name.
Tracking this as part of VCDA-3682
___
*[`pkg/capisdk/capi_gateway_manager.go` line 24 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1MIAx7FMb8owz0C3QN-r2-24:-N1srcI87Vg15z0jtMsI:b6ncvod) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/pkg/capisdk/capi_gateway_manager.go#L24)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

Can't we merge this functionality into the HTTP code itself? There is already L4 getting handled along with HTTP/HTTPS in CPI code, so potentially this function could be removed.
</blockquote></details>

There were RDE calls specific to CPI within the CPI's implementation of CreateLoadBalancer(). So I was thinking that CAPVCD specific RDE functions can come within this function. Same for UpdateLB() and DeleteLB().

I have created VCDA-3683 to follow up
___
*[`pkg/capisdk/capi_gateway_manager.go` line 28 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1MIAx7FMb8owz0C3QN-r2-28:-N1st1JiCxParon7aBC2:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/pkg/capisdk/capi_gateway_manager.go#L28)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

typo referece
</blockquote></details>

Done.
___
*[`pkg/capisdk/capi_gateway_manager.go` line 75 at r2](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/131#-N1MIAx7FMb8owz0C3QN-r2-75:-N1t-_xsEATgRhF7Bsc5:b9nzfk3) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/70a31e203a774ed3752b3e57353c758c20e4fe98/pkg/capisdk/capi_gateway_manager.go#L75)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

same comment about converting these to functions
</blockquote></details>

Added functions to convert to. generate the names

**Pull request merged**